lib/soap/request/params.ex

```diff
@@ -59,6 +66,87 @@ defmodule Soap.Request.Params do
   defp extract_headers(soap_action, []), do: base_headers(soap_action)
   defp extract_headers(_, custom_headers), do: custom_headers
 
+  @spec validate_params(params :: list(), wsdl :: map(), operation :: String.t()) :: list()
+  def validate_params(params, wsdl, operation) do
+    errors =
+      params
+      |> Enum.map(&validate_param(&1, wsdl, operation))
+
+    case Enum.any?(errors) do
+      true ->
+        {:error, Enum.reject(errors, &is_nil/1)}
+
+      _ ->
+        params
+    end
+  end
+
+  @spec validate_param(param :: tuple(), wsdl :: map(), operation :: String.t()) :: String.t() | nil
+  def validate_param(param, wsdl, operation) do
+    {k, _, v} = param
+
+    case val_map = wsdl.validation_types[String.downcase(operation)] do
+      nil ->
+        nil
+
+      _ ->
+        case Map.has_key?(val_map, k) do
```

Comment:
If...else here is better than embedded case i think

Reply:
agree with you